Overhaul Branch Logic from string based to token set based.

dannovikov opened the pull request:
### Link to Issue or Description of Change

**1. Link to an existing issue (if applicable):**

- Closes: #3470 

**Problem:**

The current approach for branching used in parallel agent for creating isolated parallel contexts for the subagents is broken and causes multiple intuitive architectures to fail. Both architectures are mentioned in the above issue. 

A sequence of parallel agents doesn't intuitively pass context; each next parallel agent's subagents don't see any of the outputs of the previous parallel groups. The linked issue above also describes another failing architecture, the fork and join, or map/reduce style architecture, which work on their own, but not when nested inside parallel agents. 

From my exploration, it was clear the string and prefix-match approach used for branching was insufficient.

**Solution:**

The proposed solution introduces a Branch class as the new branch representation, one which represents branches by sets of integer tokens.

When a ParallelAgent runs its subagents, it adds a unique token to each subagent's branch set. Then, the distinct branch sets are merged into a single set containing the union of those tokens, and this becomes the branch context going forward. 

When determining if an event is visible to a branch, we check if the event's branch token set is a subset of the current branch. A global token factory ensures each new fork gets a unique token. This token-subset approach, instead of a string prefix approach, solves the two problem architectures noted in the above issue. 

**Example:**
To explore this solution in action, consider the architecture:

Sequential[Parallel1[A, B], Parallel2[C,D], Parallel3[E, F]]

This architecture currently doesn't "work" -- each next parallel group can't see the events from the previous parallel group before them. So C, D can't see the events from A, B. 

In the newly proposed branching paradigm:

We start with an empty branch set {} and start running the sequential agent.
When Parallel1 runs, it adds unique tokens to A's and B's branch set, so A runs on branch {1} and B runs on branch {2}.

After A and B finish, the branch sets are merged, and the main context's branch is updated to the union of the parallel branches - {1,2}. 

When Parallel2 runs on branch {1, 2}, C and D are each given new branch tokens, so C runs on branch {1, 2, 3} and D runs on branch {1, 2, 4}. Both C and D can see the events that came on branch {1} and {2}, because those are subsets of each of their branches. After C and D finish, the current context's branch is updated to be their union - {1, 2, 3, 4}. 

Now {1, 2, 3, 4} is the current context, and we run Parallel 3. E gets a new token in its token set, so it runs on {1, 2, 3, 4, 5}, and F runs on {1, 2, 3, 4, 6}. They can see all the events from before; the ones that came from A on branch {1}, B on branch {2}, C on branch {1, 2, 3}, and D on branch {1, 2, 4}, each of which are subsets of the current branches. Once E and F finish, their token sets are merged, such that the context's branch going forward is {1, 2, 3, 4, 5, 6}. 

Token counter is reset on each call to runner.run_async, and the token sets are only useful for the current invocation. Events should always be visible across invocations, so if the invocation id of an event doesn't match the current invocation id, the event is always visible.

### Testing Plan

All unit tests have been updated to use the new branching logic. I have added tests to test the two architectures mentioned in the issue, and confirmed that they both work as expected under the new system.

**Unit Tests:**

- [x] I have added or updated unit tests for my change.
- [x] All unit tests pass locally.

===== 3338 passed, 2437 warnings in 30.21s =====

**Manual End-to-End (E2E) Tests:**

Please run `uv run pytest tests/unittests/agents/test_nested_agent_branch_visibility.py -vs` to see the two problem architectures running correctly and a display of how the tokens are being assigned (shown below). 

```
======================================================================
INTEGRATION TEST: Nested Parallel + Reduce (GitHub Issue #3470)
======================================================================

Architecture:
  Sequential[
    Parallel[
      Sequential[Parallel[Alice,Bob,Charlie], Reducer1],  ← Group 1
      Sequential[Parallel[David,Eve,Frank], Reducer2]     ← Group 2
    ],
    Final_Reducer  ← Sees all outputs
  ]


=== Token Distribution (Nested Parallel) ===
  Agent           Tokens                        
  --------------- ------------------------------
  Alice           | tokens=[1, 3]
  David           | tokens=[2, 6]
  Bob             | tokens=[1, 4]
  Eve             | tokens=[2, 7]
  Charlie         | tokens=[1, 5]
  Frank           | tokens=[2, 8]
  Reducer1        | tokens=[1, 3, 4, 5]
  Reducer2        | tokens=[2, 6, 7, 8]
  Final_Reducer   | tokens=[1, 2, 3, 4, 5, 6, 7, 8]
======================================================================

```
```
======================================================================
INTEGRATION TEST: Sequence of Parallels (GitHub Issue #3470)
======================================================================

Architecture:
  Sequential[
    Parallel1[Alice, Bob, Charlie],    ← Group 1
    Parallel2[David, Eve, Frank],      ← Group 2 (sees Group 1)
    Parallel3[Grace, Henry, Iris]      ← Group 3 (sees Groups 1 & 2)
  ]


=== Token Distribution ===
  Agent           Tokens                         Can See
  --------------- ------------------------------ ----------------------------------------
  --- Group 1 ---
  Alice           | tokens=[1]                          Root
  Bob             | tokens=[2]                          Root
  Charlie         | tokens=[3]                          Root
  --- Group 2 ---
  David           | tokens=[1, 2, 3, 4]                 Root, Group 1 (A,B,C)
  Eve             | tokens=[1, 2, 3, 5]                 Root, Group 1 (A,B,C)
  Frank           | tokens=[1, 2, 3, 6]                 Root, Group 1 (A,B,C)
  --- Group 3 ---
  Grace           | tokens=[1, 2, 3, 4, 5, 6, 7]        Root, Groups 1 & 2 (A-F)
  Henry           | tokens=[1, 2, 3, 4, 5, 6, 8]        Root, Groups 1 & 2 (A-F)
  Iris            | tokens=[1, 2, 3, 4, 5, 6, 9]        Root, Groups 1 & 2 (A-F)

Key Observations:
  ✓ Group 2 agents have tokens {1,2,3,...} - inherit from Group 1
  ✓ Group 3 agents have tokens {1,2,3,4,5,6,...} - inherit from Groups 1 & 2
  ✓ Each agent can see all events with token subsets
======================================================================
```



### Checklist

- [x] I have read the [CONTRIBUTING.md](https://github.com/google/adk-python/blob/main/CONTRIBUTING.md) document.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes.
- [x] I have manually tested my changes end-to-end.
- [x] Any dependent changes have been merged and published in downstream modules.

dannovikov committed:
feat(branch-context): Implement token-set provenance for parallel agent visibility

- Replace string-based branch tracking with BranchContext using frozenset tokens
- Implement fork/join semantics: parallel agents get unique tokens, join merges them
- Use subset relationship for visibility: event.tokens ⊆ agent.tokens
- Add comprehensive tests for nested parallel architectures (GitHub issue #3470)
- Add integration tests demonstrating diamond shape and loop behavior

Stats: 3321 passing tests, 17 failing (mostly serialization edge cases)

dannovikov committed:
Rename BranchContext -> Branch and update tests

dannovikov committed:
Remove extra files created during debugging

dannovikov committed:
restore accidentally deleted file

dannovikov committed:
restore accidentally deleted file

dannovikov committed:
run autoformat

dannovikov committed:
add comment and delete extra file

dannovikov committed:
make fork return one branch

dannovikov committed:
remove redundant code

dannovikov committed:
update branch types

dannovikov committed:
remove extraneous comment

dannovikov committed:
event.branch docstring update

dannovikov committed:
rename tokenfactory to BranchTokenFactory

dannovikov committed:
branch optional

dannovikov committed:
tighten up tests

dannovikov committed:
remove unused import

dannovikov committed:
tidy up

dannovikov committed:
revert changes to event converter

dannovikov committed:
tidying up

dannovikov committed:
update comments

dannovikov committed:
invocation context branch not optional

dannovikov committed:
format and rename

dannovikov committed:
Merge remote-tracking branch 'origin/main' into fix/event_filtering

dannovikov committed:
Address gemini's comments

dannovikov commented on src/google/adk/sessions/vertex_ai_session_service.py:
```diff
@@ -359,7 +359,7 @@ def _from_api_event(api_event_obj: vertexai.types.SessionEvent) -> Event:
     turn_complete = getattr(event_metadata, 'turn_complete', None)
     interrupted = getattr(event_metadata, 'interrupted', None)
-    branch = getattr(event_metadata, 'branch', None)
+    branch = getattr(event_metadata, 'branch', None) or None
```
the Vertex AI API wouldn't return a Branch object directly, right? So I've condensed your suggestion to just:

```
branch_raw = getattr(event_metadata, 'branch', None)
    branch: Optional[Branch] = None
    if isinstance(branch_raw, dict):
      branch = Branch.model_validate(branch_raw)
```

dannovikov replied:
/gemini review

dannovikov committed:
Address gemini's comments 2

ryanaiagent commented:
Hi @dannovikov , Thank you for your contribution! We appreciate you taking the time to submit this pull request.
Can you please fix the failing tests

dannovikov committed:
fix failing unit test missing import

dannovikov committed:
Merge branch 'main' into develop

dannovikov committed:
feat: structural branch filtering with O(D) memory complexity

dannovikov committed:
Revert "feat: structural branch filtering with O(D) memory complexity"

This reverts commit de75ffadd5c09a2976d225d493a2faa74a53f60d.

dannovikov committed:
fix(tests): Update test fixtures for token-based Branch model

- Update test_interactions_processor.py to use Branch objects instead of
  strings, aligning with the new token-based branch model
- Add missing mock_bq_client fixture to test_global_client_reuse to prevent
  hanging when creating real BigQuery clients

dannovikov committed:
autoformat
